### PR TITLE
Fix and improvement of the Scaffold module

### PR DESCRIPTION
**Pull request by crazycat256**
## Type of change

- [x] Bug fix
- [x] New feature

## Description

- Fixed an issue where the non-air-place mode would cause the player to fall when descending.
- Fixed a bug in the air-place mode.
- Improved the logic for placing blocks in the non-air-place mode.
- Enhanced the fast-tower functionality.
- Added the "ahead-distance" setting, which allows choosing the distance at which blocks are placed in front of the player (prevent falling in certain cases).

## Related issues
[[Bug] Scaffold does not work #3543](https://github.com/MeteorDevelopment/meteor-client/issues/3543)

# How Has This Been Tested?

:)

# Checklist:

- [x] My code follows the style guidelines of this project.
- [x] I have added comments to my code in more complex areas.
- [x] I have tested the code in both development and production environments.

### CODE REVIEW

File: src/main/java/meteordevelopment/meteorclient/systems/modules/movement/Scaffold.java [hunk 1]
```diff
@@ -259,9 +253,23 @@ private void onTick(TickEvent.Pre event) {
...
-            if (!mc.world.getBlockState(bp).isAir()) {
-                prevBp.set(bp);
+            place(bp);
+        }
+
+        if (fastTower.get() && mc.options.jumpKey.isPressed() && !mc.options.sneakKey.isPressed() && InvUtils.findInHotbar(itemStack -> validItem(itemStack, bp)).found()) {
+            Vec3d velocity = mc.player.getVelocity();
+            Box playerBox = mc.player.getBoundingBox();
+            if (Streams.stream(mc.world.getBlockCollisions(mc.player, playerBox.offset(0, 1, 0))).toList().isEmpty() /*||
+                !Streams.stream(mc.world.getBlockCollisions(mc.player, playerBox.offset(0, -1, 0))).toList().isEmpty()*/) {
```
Comment (Wide-Cat) [inline src/main/java/meteordevelopment/meteorclient/systems/modules/movement/Scaffold.java:263]: 👁️

File: src/main/java/meteordevelopment/meteorclient/systems/modules/movement/Scaffold.java [hunk 2]
```diff
@@ -159,88 +170,72 @@ public class Scaffold extends Module {
...
-    public void onActivate() {
-        lastWasSneaking = mc.options.sneakKey.isPressed();
-        if (lastWasSneaking) lastSneakingY = mc.player.getY();
-    }
 
     @EventHandler
     private void onTick(TickEvent.Pre event) {
         if (onlyOnClick.get() && !mc.options.useKey.isPressed()) return;
 
+        Vec3d vec = mc.player.getPos().add(mc.player.getVelocity()).add(0, -0.75, 0);
```
Comment (Wide-Cat) [inline src/main/java/meteordevelopment/meteorclient/systems/modules/movement/Scaffold.java:183]: is 0.75 for gravity or something?

**Review by Wide-Cat — CHANGES_REQUESTED** [review]

File: src/main/java/meteordevelopment/meteorclient/systems/modules/movement/Scaffold.java (hunk 2 above)
Comment (crazycat256) [reply src/main/java/meteordevelopment/meteorclient/systems/modules/movement/Scaffold.java:183]: This value determines how far below the player the blocks are placed, so logically it should be the smallest value greater than -1 (-0.999...).
However, setting it to -0.75 avoids placing blocks under the bridge like this:

https://github.com/MeteorDevelopment/meteor-client/assets/63475640/eb6e9f40-4592-40cb-ba4d-cf62742f1489

**Comment by Wide-Cat** [comment]
It seems to be quite buggy if you have fast tower enabled and no blocks available to place. If you have no blocks at all in your hotbar you seem to be prevented from jumping when you stand still and it tries to tower up, while if you do have blocks in your hotbar but you have autoswitch disabled and you arent holding a stack, you go up but dont place any blocks at all. It would also be nice to have a speed setting for fast tower.

**Comment by crazycat256** [comment]
@Wide-Cat I've fixed the bugs you describe and added a setting to control the speed of the fast-tower.

**Review by Wide-Cat — APPROVED** [review]

**Pull request merged**